File: frontend/src/utils/attendanceStats.js
// Haqiqiy /api/attendance/student/{id} hodisalaridan (IN/OUT) statistikani hisoblaydi.
// MUHIM: "kechikish" (late) tushunchasi backendda yo'q (maktab boshlanish vaqti schema'da
// mavjud emas — V1ReportsController bilan bir xil qoida), shuning uchun bu yerda ham
// hech qanday "late" son o'ylab topilmaydi — faqat mavjud IN/OUT hodisalaridan kelib
// chiqadigan haqiqiy sonlar qaytariladi.

function dateKey(d) {
  return d.toISOString().slice(0, 10);
}

// Dushanba-Shanba maktab kuni deb hisoblanadi (Yakshanba dam olish kuni), boshqa joylarda
// (V1ReportsController haftalik hisobot) ishlatilgan konventsiyaga mos.
function isSchoolDay(d) {
  return d.getDay() !== 0;
}

export function computeAttendanceStats(events, days = 30) {
  const now = new Date();
  const from = new Date(now);
  from.setDate(from.getDate() - (days - 1));
  from.setHours(0, 0, 0, 0);

  const presentDates = new Set();
  for (const ev of events || []) {
    if (ev.type !== 'IN') continue;
    const ts = new Date(ev.timestamp);
    if (ts < from || ts > now) continue;
    presentDates.add(dateKey(ts));
  }

  let schoolDays = 0;
  for (let d = new Date(from); d <= now; d.setDate(d.getDate() + 1)) {
    if (isSchoolDay(d)) schoolDays++;
  }

  const presentDays = presentDates.size;
  const absentDays = Math.max(0, schoolDays - presentDays);
  const percent = schoolDays > 0 ? Math.round((presentDays / schoolDays) * 100) : 0;

  return { presentDays, absentDays, schoolDays, percent, periodDays: days };
}

// Faqat haqiqiy sonlardan tuzilgan qisqa xulosa jumlasi — shaxsiyat/psixologik
// bahо EMAS, faqat davomat foizi va kunlar sonini tavsiflaydi.
export function attendanceSummaryKey(stats) {
  if (stats.schoolDays === 0) return 'noData';
  if (stats.presentDays === 0) return 'noAttendance';
  if (stats.percent >= 90) return 'excellent';
  if (stats.percent >= 75) return 'good';
  if (stats.percent >= 50) return 'moderate';
  return 'low';
}

File: mobile/lib/core/attendance_stats.dart
import '../models/models.dart';

/// Veb-ilova (frontend/src/utils/attendanceStats.js) bilan bir xil hisob-kitob —
/// faqat haqiqiy IN/OUT hodisalaridan. "Kechikish" hisoblanmaydi, chunki backend
/// schemasida maktab boshlanish vaqti tushunchasi yo'q (V1ReportsController bilan
/// bir xil qoida).
class AttendanceStats {
  final int presentDays;
  final int absentDays;
  final int schoolDays;
  final int percent;
  final int periodDays;

  AttendanceStats({
    required this.presentDays,
    required this.absentDays,
    required this.schoolDays,
    required this.percent,
    required this.periodDays,
  });
}

bool _isSchoolDay(DateTime d) => d.weekday != DateTime.sunday;

AttendanceStats computeAttendanceStats(List<AttendanceEvent> events, {int days = 30}) {
  final now = DateTime.now();
  final from = DateTime(now.year, now.month, now.day).subtract(Duration(days: days - 1));

  final presentDates = <String>{};
  for (final ev in events) {
    if (ev.type != 'IN') continue;
    if (ev.timestamp.isBefore(from) || ev.timestamp.isAfter(now)) continue;
    presentDates.add('${ev.timestamp.year}-${ev.timestamp.month}-${ev.timestamp.day}');
  }

  var schoolDays = 0;
  for (var d = from; !d.isAfter(now); d = d.add(const Duration(days: 1))) {
    if (_isSchoolDay(d)) schoolDays++;
  }

  final presentDays = presentDates.length;
  final absentDays = (schoolDays - presentDays).clamp(0, schoolDays);
  final percent = schoolDays > 0 ? ((presentDays / schoolDays) * 100).round() : 0;

  return AttendanceStats(
    presentDays: presentDays,
    absentDays: absentDays,
    schoolDays: schoolDays,
    percent: percent,
    periodDays: days,
  );
}

/// Faqat haqiqiy sonlardan tuzilgan xulosa kaliti — shaxsiyat/psixologik baho EMAS.
String attendanceSummaryKey(AttendanceStats s) {
  if (s.schoolDays == 0) return 'noData';
  if (s.presentDays == 0) return 'noAttendance';
  if (s.percent >= 90) return 'excellent';
  if (s.percent >= 75) return 'good';
  if (s.percent >= 50) return 'moderate';
  return 'low';
}
